### PR TITLE
Optional baseLog

### PR DESCRIPTION
**Pull request by robmcguinness**
This is essentially #282 with requested changes. Please review and provide feedback

### CODE REVIEW

File: pino.js [hunk 1]
```diff
@@ -338,6 +344,18 @@ function pino (opts, stream) {
...
 
+  var base = (typeof iopts.base === 'object') ? iopts.base : defaultBase
+
+  if (iopts.name !== undefined) {
+    base = Object.assign({}, base, {
+      name: iopts.name
+    })
+  }
+
+  if (base != null) {
```
Comment (jsumners) [inline pino.js]: Is there a reason for the `!=` as apposed to `!==` or `(!base)`?

File: pino.js [hunk 2]
```diff
@@ -36,6 +37,11 @@ var defaultOptions = {
   messageKey: 'msg'
 }
 
+var defaultBase = {
```
Comment (jsumners) [inline pino.js]: Why isn't this part of the `defaultOptions` object?

File: pino.js [hunk 3]
```diff
@@ -338,6 +344,18 @@ function pino (opts, stream) {
...
+  var base = (typeof iopts.base === 'object') ? iopts.base : defaultBase
+
+  if (iopts.name !== undefined) {
+    base = Object.assign({}, base, {
+      name: iopts.name
+    })
+  }
+
+  if (base != null) {
+    instance = instance.child(base)
```
Comment (jsumners) [inline pino.js]: The default instance should not be a child.

cc: @davidmarkclements @mcollina

File: pino.js (hunk 1 above)
Comment (robmcguinness) [reply pino.js]: no reason my mistake

File: pino.js (hunk 2 above)
Comment (robmcguinness) [reply pino.js]: I simply renamed `defaultBase` according to the comments in #282 (from a previously submitted PR). I don't see a reason why the `base` object can't be a part of `defaultOptions`. Would you rather have something like:

File: pino.js (hunk 3 above)
Comment (robmcguinness) [reply pino.js]: I was following the comments from https://github.com/pinojs/pino/pull/282#pullrequestreview-57188442. Unfortunately I don't know enough about Pino to understand the implications of using a child logger for the default instance. When I get a moment I'll dig deeper into Pino.

File: pino.js (hunk 2 above)
Comment (jsumners) [reply pino.js]: Yes, please.

File: pino.js (hunk 3 above)
Comment (jsumners) [reply pino.js]: It seems odd to me. But that's why I pinged the other two, so they can provide enlightenment.

Comment (mcollina) [reply pino.js]: why not? I think it's the best way to do it. Considering that pino children do not keep a reference to their parent, I see no obvious problem with this.

File: pino.js [hunk 4]
```diff
@@ -128,7 +133,7 @@ function asJson (obj, msg, num) {
   var hasObj = obj !== undefined && obj !== null
   var objError = hasObj && obj instanceof Error
   msg = !msg && objError ? obj.message : msg || undefined
-  var data = this._baseLog + this._lscache[num] + this.time()
+  var data = '{' + this._lscache[num] + this.time()
```
Comment (mcollina) [inline pino.js]: can you embed `{` into `this._lscache`?

File: pino.js (hunk 3 above)
Comment (mcollina) [reply pino.js]: The only side effect of this is that it makes the base-log properties overridable by a log line. I don't think it's a bad thing: https://github.com/pinojs/pino/blob/master/pino.js#L136-L138.

Comment (jsumners) [reply pino.js]: Okay.

File: pino.js (hunk 4 above)
Comment (robmcguinness) [reply pino.js]: @mcollina not sure if this is what you wanted:

Comment (mcollina) [reply pino.js]: ??

Comment (robmcguinness) [reply pino.js]: You asked: 

> can you embed { into this._lscache?

So I provided some examples above and what I thought that meant. Apologies if I misinterpreted the ask.

Comment (robmcguinness) [reply pino.js]: added one more commit. let me know if that is what you were requesting.

**Comment by mcollina** [comment]
Folks, weight in if this is semver-major or not.

**Comment by jsumners** [comment]
It isn't. The default operation of the API hasn't changed from a user perspective. It would merely be possible to do more things.

**Comment by robmcguinness** [comment]
conflicts should be resolved now

**Comment by mcollina** [comment]
@jsumners @davidmarkclements may I get a couple of LGTMs?

**Comment by jsumners** [comment]
Sorry. Thought I already had.

**Pull request merged**